csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -301,27 +133,257 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
+    // 2. The extent of sharded id in producer / consumer is divisible by
+    // split_factor. NOTE: We can only check if DID(d) is on the outer of the
+    // split regardless of the split_factor. However, when applying the split to
+    // the target, the split_factor will need to be num_devices. For e.g.: A[h]
+    // -> reshape -> B[a, h/a] If A is inner split `h/d`, then directly
+    // replaying the split on `a` will produce `a/(h/d), h/d` instead of `d,
+    // a/d`. So we should instead outer split by num_devices.
+
+    // Find the consumer between the reference and target.
+    auto [consumer_id, consumer_tv] = direction == PropagateDirection::kForward
+        ? std::make_pair(target_id, target)
+        : std::make_pair(ref_id, ref);
+
+    if (hasRootToLogicalTransform(consumer_id, consumer_tv)) {
+      validate_split(split, target_id);
+    }
+
+    auto it = transformed_loop.erase(target_id).second;
+    auto [outer, inner] =
+        IterDomain::split(target_id, split->factor(), split->innerSplit());
+
+    transformed_loop.insert(it, outer, std::monostate());
+    transformed_loop.insert(it, inner, std::monostate());
+
+    // Add mapping between ref and target for the propagated DID split.
+    // This is used to propagate 2D sharding and parallelization.
+    ref2target[split->outer()] = outer;
+    ref2target[split->inner()] = inner;
+  }
+
+  // Parallelize based on the ref2target map.
+  for (IterDomain* device_id : device_ids) {
+    NVF_ERROR(
+        ref2target.find(device_id) != ref2target.end(),
+        "Failed to propagate ",
+        device_id,
+        " to ",
+        target);
+    IterDomain* target_id = ref2target.at(device_id);
+    target_id->parallelize(device_id->getParallelType());
+  }
+
+  auto new_loop = std::views::keys(transformed_loop);
+  target->setLoopDomain({new_loop.begin(), new_loop.end()});
+}
+
+void propagateDIDTransform(
+    const TensorView* ref,
+    TensorView* tv,
+    const std::unordered_set<ParallelType>& selected_parallel_types,
+    PropagateDirection direction) {
+  tv->setDeviceMesh(ref->getDeviceMesh());
+
+  std::unordered_set<IterDomain*> device_ids;
+  const std::unordered_map<IterDomain*, IterDomain*> ref2target =
+      getRef2TargetMap(ref, tv, direction);
+
+  for (IterDomain* maybe_did : ref->getLoopDomain()) {
+    if (selected_parallel_types.count(maybe_did->getParallelType()) == 0) {
+      continue;
+    }
+    // Get input of maybe_did in the root / logical domain
+    // that will be present in ref2target mapping.
+    std::unordered_set<IterDomain*> inputs =
+        direction == PropagateDirection::kForward
+        ? getInputsInTargetDomain({maybe_did}, ref->getLogicalDomain())
+        : getInputsInTargetDomain({maybe_did}, ref->getMaybeRootDomain());
+    NVF_ERROR(
+        inputs.size() == 1,
+        "Expected one input for ",
+        maybe_did,
+        " in the root / logical domain. Got ",
+        inputs.size(),
+        " inputs.");
```

Comment:
```suggestion
    NVF_ERROR_EQ(
        inputs.size(), 1,
        "Expected one input for ",
        maybe_did,
        " in the root / logical domain.");
```
NVF_*_EQ is designed for printing runtime values so you don't have to manually print them.